### PR TITLE
Add `device_matrix_data` and device-side `matrix::read`

### LINKED ISSUE #729: Repeated allocations while reading `Fbcsr` matrix
Currently, the `Fbcsr` `read` function allocates one `std::vector` for every small block. This may impact read performance for the smallest block sizes.

Since `read` is typically called very few times per program execution, this is low-priority.

Related pull-request: https://github.com/ginkgo-project/ginkgo/pull/671

### PR DESCRIPTION
**Pull request by upsj**
This PR adds a new `device_matrix_data` that works equivalently to `matrix_data` except that it uses a device-side `Array`. I provide a default implementation using the old host-side initialization.

There are still some things to do, but I wanted to ask for comments already.

- [x] merge #833 to implement Ell `compute_max_nnz`
- [x] add `Ell::read`
- [x] add `Hybrid::read`
- [x] add `Sellp::read`
- [x] ~~modify Ell/Sellp sentinel values~~ #905
- [x] ~~modify conversions for numerical zeros~~ This will be taken care of with #905 
- [x] ~~move `read` tests from `core` to `reference`~~ (this will no longer be necessary with #889)
- [x] ~~add tests checking zero behavior~~ This will be taken care of with #904
- [x] add tests for `remove_zeros` and device_matrix_data functionality

Fixes #729

### CODE REVIEW

**Review by yhmtsai — COMMENTED** [review]
First part of my review.
It does not contain hybrid/sellp/fb_csr implementation.
Some comments containing Reminder is a note for me because it is hard to go different file on GitHub review.

File: common/unified/matrix/ell_kernels.cpp
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:90]: this workaround is only necessary for structs containing std/thrust::complex members, since its alignment restrictions cause the struct to have a different size between GPU and CPU. So this is specific to members, not values.

File: common/unified/matrix/ell_kernels.cpp
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:95]: The function takes a reference, so the values doesn't actually need to be read. Another solution could be `zero<device_type<ValueType>>()`, I guess?

File: common/unified/matrix/ell_kernels.cpp
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:94]: That will change in another PR, I wanted to keep the fundamental changes out of this PR

File: core/base/device_matrix_data.cpp
Comment (upsj) [reply core/base/device_matrix_data.cpp:66]: I am trying to reduce the amount of function implementations we have in our headers, since they make it harder to find important things, and make the compilation more costly.

File: core/base/device_matrix_data.cpp
Comment (upsj) [reply core/base/device_matrix_data.cpp:90]: I'm trying to keep the overhead of create_from_host minimal when exec is a CPU executor. But you make a good point - it should probably have `view` in its name, or remove the `std::move` call in the constructor of the resulting object to copy the data.

File: core/matrix/coo.cpp
Comment (upsj) [reply core/matrix/coo.cpp:217]: I don't think that is worth it, `write` performance seems much less necessary to me than `read` performance.

File: core/matrix/dense.cpp
Comment (upsj) [reply core/matrix/dense.cpp:679]: My intention here was to avoid reallocations if possible in all cases. That makes interaction with array views much simpler

File: core/matrix/sparsity_csr.cpp
Comment (upsj) [reply core/matrix/sparsity_csr.cpp:109]: That's the current behavior for read, IIRC

File: dpcpp/CMakeLists.txt
Comment (upsj) [reply dpcpp/CMakeLists.txt:58]: I think we should wait for Alina's feedback when the oneDPL folks got back to her.

File: include/ginkgo/core/base/range.hpp
Comment (upsj) [reply include/ginkgo/core/base/range.hpp:111]: adding the assertion here breaks ROCm compilation due to some weird include order issue after adding the device_matrix_data include to types.hpp

File: include/ginkgo/core/matrix/fbcsr.hpp
Comment (upsj) [reply include/ginkgo/core/matrix/fbcsr.hpp:311]: size is always divisible by block size

File: include/ginkgo/core/matrix/fbcsr.hpp
Comment (upsj) [reply include/ginkgo/core/matrix/fbcsr.hpp:308]: Adding this function was a mistake, as it allows users to completely break the Fbcsr type. Any use on a non-empty matrix causes any subsequent operations to either produce wrong results or (more likely) segfault.

File: include/ginkgo/core/matrix/sellp.hpp
Comment (upsj) [reply include/ginkgo/core/matrix/sellp.hpp:285]: yep, I'll try to revert that before merging.

File: include/ginkgo/core/matrix/sellp.hpp
Comment (upsj) [reply include/ginkgo/core/matrix/sellp.hpp:332]: total_cols can be computed from the size of `values` and `col_idxs`, so we don't need to store it.

File: core/matrix/diagonal.cpp
Comment (upsj) [reply core/matrix/diagonal.cpp:254]: As we don't have dense::convert_to(diagonal), so I would keep these operations conceptually separate.

File: common/unified/matrix/ell_kernels.cpp
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp:93]: right, I did the same thing for Sellp, will do that.

File: common/unified/matrix/ell_kernels.cpp
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp:95]: > The function takes a reference, so the values doesn't actually need to be read. 

Yes, I think so, too. the function only uses the type of value to decide which one to run.
> Another solution could be `zero<device_type<ValueType>>()`, I guess?

Yes

File: common/unified/matrix/ell_kernels.cpp
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp:94]: I see.

File: core/base/device_matrix_data.cpp
Comment (yhmtsai) [reply core/base/device_matrix_data.cpp:90]: after reviewing more files, I realize the executor can also be host, so it is view actually when executor is host

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc
Comment (thoasm) [inline common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Shouldn't this be `num_rows + 1`?
```suggestion
        components::fill_array(exec, row_ptrs, num_rows + 1, IndexType{});
```

**Review by thoasm — APPROVED** [review]
Mostly nits, so LGTM!

File: cuda/base/types.hpp
Comment (upsj) [reply cuda/base/types.hpp:407]: yes, though with using as_cuda_type, that means effectively the same thing.

File: common/unified/matrix/sparsity_csr_kernels.cpp
Comment (upsj) [reply common/unified/matrix/sparsity_csr_kernels.cpp:51]: how about using fill_in_matrix_data consistently everywhere? The issue is really that each matrix format is slightly different, so I tried to use a consistent name for the part that actually finishes the matrix data read, except for Diagonal and Dense, where it just fills in the non-zeros into an already zeroed array.

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Yep, due to all initializations going through Reference, we aren't actually testing the device-side read functions right now. Working on fixing that.

File: cuda/base/types.hpp
Comment (upsj) [reply cuda/base/types.hpp:69]: Hope that doesn't break CUDA 😄

File: common/unified/matrix/sparsity_csr_kernels.cpp
Comment (thoasm) [reply common/unified/matrix/sparsity_csr_kernels.cpp:51]: Sounds good

File: reference/matrix/ell_kernels.cpp
Comment (upsj) [reply reference/matrix/ell_kernels.cpp:173]: I'd prefer to keep the loops simple/unsurprising, separate increments might get lost when reading the code.

File: reference/matrix/diagonal_kernels.cpp
Comment (upsj) [reply reference/matrix/diagonal_kernels.cpp:144]: It sounds like you're describing invalid matrix_data with out-of-bounds entries - should that be supported/caught? I'd probably defer that to device_matrix_data::validate() and #770

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:92]: Actually, I'll add tests and potential fixes in #904

File: core/matrix/diagonal.cpp
Comment (yhmtsai) [reply core/matrix/diagonal.cpp:254]: I use the wrong function. it is the extract_diagonal

File: core/matrix/sparsity_csr.cpp
Comment (yhmtsai) [reply core/matrix/sparsity_csr.cpp:109]: Yes, I am thinking whether using the first value as the diagonal makes sense or not.
let's leave it as the original behavior now

File: include/ginkgo/core/matrix/fbcsr.hpp
Comment (yhmtsai) [reply include/ginkgo/core/matrix/fbcsr.hpp:308]: I see. it is a bug-fix, so we can change it

File: reference/matrix/diagonal_kernels.cpp
Comment (yhmtsai) [reply reference/matrix/diagonal_kernels.cpp:144]: Maybe? it something like you did for core/test/utils/matrix_generator.hpp?

Comment (upsj) [reply reference/matrix/diagonal_kernels.cpp:144]: That was a bugfix for something that was broken earlier. To be honest, I think we should not have added a `read` function to Diagonal at all, since it is not a generic matrix format. But that was an earlier oversight, we can't change that until 2.0. Until then, I'd like to avoid having deal with device-side error handling

**Review by yhmtsai — APPROVED** [review]
LGTM. only some nit/comments

File: common/unified/matrix/ell_kernels.cpp
Comment (yhmtsai) [reply common/unified/matrix/ell_kernels.cpp]: is unpack_member the rename of fake_complex_unpack? Is there any difference between kernel_launch and type?

File: core/matrix/dense.cpp
Comment (yhmtsai) [reply core/matrix/dense.cpp:679]: what do you mean array view here? it does not use view here, I think?
In current develop, we do not adapt the input, either. Thus, it is fine now.
relative issue #96

Comment (upsj) [reply core/matrix/dense.cpp:679]: If the matrix already contains data in a view with a stride, this approach reuses the storage, which is probably closer to what users might expect?

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:126]: it's the starting index of the block

File: omp/matrix/fbcsr_kernels.cpp
Comment (upsj) [reply omp/matrix/fbcsr_kernels.cpp:193]: Unfortunately, it wasn't entirely equivalent, I use the correct one in both places now

File: common/cuda_hip/matrix/fbcsr_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/matrix/fbcsr_kernels.hpp.inc:148]: You are right, it should be num_blocks - 1

File: common/unified/matrix/ell_kernels.cpp
Comment (upsj) [reply common/unified/matrix/ell_kernels.cpp]: One is device-specific (only for CUDA/HIP), the other one is generic in common kernels.

File: core/matrix/diagonal.cpp
Comment (upsj) [reply core/matrix/diagonal.cpp:254]: I can't follow. A diagonal matrix is always square, so what would be the issue here?

File: dpcpp/components/device_matrix_data_kernels.dp.cpp
Comment (upsj) [reply dpcpp/components/device_matrix_data_kernels.dp.cpp:36]: Good point, I forced its location now for the format_header script as well

File: common/cuda_hip/matrix/hybrid_kernels.hpp.inc
Comment (upsj) [reply common/cuda_hip/matrix/hybrid_kernels.hpp.inc:146]: This is only necessary because a Lambda wouldn't be copy-assignable, which one of the algorithms requires. So this is only a local workaround without any generic usage.

File: core/matrix/hybrid.cpp
Comment (upsj) [reply core/matrix/hybrid.cpp:207]: That is true, though I would separate between row pointers (indexing nonzeros, why I'm using int64 here) and row indices (indexing rows) here. The former are more likely to overflow than the latter.

File: core/matrix/diagonal.cpp
Comment (yhmtsai) [reply core/matrix/diagonal.cpp:254]: extract_diagonal works for non-square matrix. do we also use this rule for reading?

**Pull request merged**
